utils/fs/memory/memory.go

```diff
@@ -0,0 +1,333 @@
+package memory
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"gopkg.in/src-d/go-git.v4/utils/fs"
+)
+
+const separator = '/'
+
+// Memory a very convenient filesystem based on memory files
+type Memory struct {
+	base string
+	s    *storage
+}
+
+//New returns a new Memory filesystem
+func New() *Memory {
+	return &Memory{
+		base: "/",
+		s:    &storage{make(map[string]*file, 0)},
+	}
+}
+
+func (fs *Memory) Create(filename string) (fs.File, error) {
+	return fs.OpenFile(filename, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0666)
+}
+
+func (fs *Memory) Open(filename string) (fs.File, error) {
+	return fs.OpenFile(filename, os.O_RDONLY, 0)
+}
+
+func (fs *Memory) OpenFile(filename string, flag int, perm os.FileMode) (fs.File, error) {
+	fullpath := fs.Join(fs.base, filename)
+	f, ok := fs.s.files[fullpath]
+	if !ok && flag&os.O_CREATE == 0 {
```

Comment:
I really love this kind of abstractions, but this is more a general problem, not only related with this package.